### PR TITLE
Fix contact category association

### LINKED ISSUE #39346: Associations not working for com_contact in Language Switcher module
### Steps to reproduce the issue

Set up the Joomla instance as multilingual, eg French and English, and ensure the Language Switcher module is published. 
Define language-specific com_contact categories – eg 'cat-fr' for French and 'cat-en' for English
Define associations to link those categories.
Define language-specific menus, and include a menuitem on each menu pointing to a menuitem of type
Contacts / List Contacts in a Category, and select the category for that language.

So you have something like
- English menu, with English menuitem pointing to English contact category 'cat-en'
- French menu, with French menuitem pointing to French contact category 'cat-fr'
- The English and French categories set as Associations of each other.

On the front-end click on eg the English menuitem to show the English category.
Then click on the French flag in the language switcher.

### Expected result

You get routed to the French category equivalent.

### Actual result

You stay on the English category page.

### System information (as much as possible)

Joomla v4.2.5

### Additional comments

I'm pretty sure that the problem is that in `administrator\components\com_categories\src\Helper\CategoryAssociationHelper.php`, in `getCategoryAssociations` it has

    $helperClassname = ucfirst(substr($extension, 4)) . 'HelperRoute';
    …
    if (class_exists($helperClassname) && \is_callable(array($helperClassname, 'getCategoryRoute'))) {
          $return[$tag] = $helperClassname::getCategoryRoute($item, $tag, $layout);
    } else {
    …

The `$helperClassname` is not being found and so the com_contact `getCategoryRoute` method is not called.

The same problem doesn't occur for com_content because it has in `libraries\extensions.classmap.php` the line:

    JLoader::registerAlias('ContentHelperRoute', '\\Joomla\\Component\\Content\\Site\\Helper\\RouteHelper', '5.0');

When I inserted a similar line into that file for com_contact:

    JLoader::registerAlias('ContactHelperRoute', '\\Joomla\\Component\\Contact\\Site\\Helper\\RouteHelper', '5.0');

it all worked ok.

For comparison, the `getCategoryRoute` method is also called from `plugins\finder\categories\categories.php`, in the `index` function, (where it looks for the method in both the `RouterHelper` class and the `helpers/route.php` file), and from `getItemRoute` in `components\com_tags\src\Helper\RouteHelper.php`. 

It would make sense to have a common approach for finding this `getCategoryRoute` method. 

Note also that this problem probably occurs with other Joomla components which support categories (excepting com_content).

See also [Improve how ContentHelperRoute is registered](https://issues.joomla.org/tracker/joomla-cms/25203) .

### PR DESCRIPTION
**Pull request by joomdonation**
Fixes #39346.

### Summary of Changes
This PR fixes wrong link generated by Language Switcher module for contact category association. Please look at https://github.com/joomla/joomla-cms/issues/39346 to understand the issue.

### Testing Instructions
1. Setup a sample multilingual website using Joomla 4.2
2. Follow the instructions at https://github.com/joomla/joomla-cms/issues/39346 , confirm the issue.
3. Apply patch, confirm the issue fixed.

### Link to documentations
Please select:
- [x] No documentation changes for docs.joomla.org needed

- [x] No documentation changes for manual.joomla.org needed

### CODE REVIEW

File: administrator/components/com_categories/src/Helper/CategoryAssociationHelper.php [hunk 1]
```diff
@@ -53,9 +53,17 @@ public static function getCategoryAssociations($id = 0, $extension = 'com_conten
...
-                    $return[$tag] = 'index.php?option=' . $extension . '&view=category&id=' . $item . $viewLayout;
+                    if ($tag && $tag !== '*') {
+                        $link .= '&lang=' . $tag;
+                    }
+
+                    if ($layout) {
+                        $link .= '&layout=' . $layout;
+                    }
+                    
+                    $return[$tag] = $link;
```
Comment (richard67) [inline administrator/components/com_categories/src/Helper/CategoryAssociationHelper.php]:
```suggestion
                    }

                    $return[$tag] = $link;
```
Fix PHPCS in Drone.

Comment (joomdonation) [reply administrator/components/com_categories/src/Helper/CategoryAssociationHelper.php]: Thanks Richard.

**Comment by brianteeman** [comment]
I have tested this item :white_check_mark: successfully on 534b67e8de0a97fbf1c174c46885fb46a4f64d1f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39358">issues.joomla.org/tracker/joomla-cms/39358</a>.</sub>

**Review by richard67 — APPROVED** [review]

**Comment by viocassel** [comment]
I have tested this item :white_check_mark: successfully on 534b67e8de0a97fbf1c174c46885fb46a4f64d1f<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39358">issues.joomla.org/tracker/joomla-cms/39358</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/39358">issues.joomla.org/tracker/joomla-cms/39358</a>.</sub>

**Comment by robbiejackson** [comment]
I couldn't find a Test This button, but I confirm that this fixes the issue 39346 I raised, and works well too as a solution for 3rd party components. In the case of the Joomla MVC tutorial all that the override of the `getCategoryRoute` method did here was add in the `&lang=` parameter into the query. 

Just as an incidental query, is there a particular reason why this `CategoryAssociationHelper` class is marked as abstract?

**Comment by joomdonation** [comment]
> Just as an incidental query, is there a particular reason why this `CategoryAssociationHelper` class is marked as abstract?

I guess it is defined like that to prevent us from instantiate object of that class. It only contains one static method, so there is no reason to create object of that class (I'm unsure if that's the right reason).

**Comment by roland-d** [comment]
Thank you.

**Pull request merged**